connectors/sources/google_drive.py

```diff
@@ -195,22 +206,234 @@ async def _execute_api_call(self, resource, method, call_api_func, kwargs):
+        """Retrieves all folders from Google Drive
+
+        Returns:
+            dict: mapping between folder id and its (name, parents)
+        """
+        folders = {}
+        async for page in self.get_folders():
+            folders_chunk = page.get("files", [])
+            for folder in folders_chunk:
+                folders[folder["id"]] = {
+                    "name": folder["name"],
+                    "parents": folder.get("parents", None),
+                }
+
+        return folders
+
+    async def list_files(self):
+        """Get files from Google Drive. Files can have any type.
+
+        Yields:
+            dict: Documents from Google Drive.
+        """
+        async for file in self.api_call_paged(
+            resource="files",
+            method="list",
+            corpora="allDrives",
+            q="trashed=false",
+            orderBy="modifiedTime desc",
+            fields=f"files({DRIVE_ITEMS_FIELDS}),incompleteSearch,nextPageToken",
+            includeItemsFromAllDrives=True,
+            supportsAllDrives=True,
+            pageSize=100,
+        ):
+            yield file
+
+    async def list_permissions(self, file_id):
+        """Get permissions for a given file ID from Google Drive.
+
+        Yields:
+            dictionary: Permissions from Google Drive for a file.
+        """
+        async for permission in self.api_call_paged(
+            resource="permissions",
+            method="list",
+            fileId=file_id,
+            fields="permissions(type,emailAddress,domain),nextPageToken",
+            supportsAllDrives=True,
+            pageSize=100,
+        ):
+            yield permission
+
+
+class GoogleAdminDirectoryClient(GoogleAPIClient):
+    """A google admin directory client to handle api calls made to Google Admin API."""
+
+    def __init__(self, json_credentials, subject):
+        """Initialize the GoogleApiClient superclass.
+
+        Args:
+            json_credentials (dict): Service account credentials json.
+            subject (str): For service accounts with domain-wide delegation enabled. A user
+                           account to impersonate - e.g "admin@your-organization.com"
+        """
+        super().__init__(
+            json_credentials=json_credentials,
+            api_name="admin",
+            api_version="directory_v1",
+            scopes=[
+                "https://www.googleapis.com/auth/admin.directory.group.readonly",
+                "https://www.googleapis.com/auth/admin.directory.user.readonly",
+            ],
+            subject=subject,
+        )
+        self.domain = _get_domain_from_email(subject)
+
+    async def list_users(self):
+        """Get files from Google Drive. Files can have any type.
+
+        Yields:
+            dict: Documents from Google Drive.
+        """
+        async for user in self.api_call_paged(
+            resource="users",
+            method="list",
+            domain=self.domain,
```

Comment:
It seems you only try to get the `domain` from the admin email, instead of impersonating it. In this case, can we just ask customers to provide the domain instead of the admin user email?

Additional question, is it possible to get domain from the service account?

Reply:
For that api call we explicitly need a domain (or customer id, [api documentation](https://developers.google.com/admin-sdk/directory/reference/rest/v1/users/list)). 

Note that we need to use google workspace admin email in:
`class GoogleAdminDirectoryClient(GoogleAPIClient):` constructor (it's called `subject` by the google API client). 

Service accounts are never tied to workspaces (domains), therefore we need to perform domain-wide delegation and impersonate admin with minimal permission scopes to fetch all users.